### DISCUSSION TITLE
Ethlib Demo Setup

### DISCUSSION THREAD

**Discussion by epsilon-0311** [t-crest/patmos#96] [Q&A] [answered]
Hi all,

I'm trying getting used to Patmos and I wanted to check out how the ethlib works. Unfortunately I can't get it working on my DE2-115 Board.

I selected the altde2-115 config get the Eth Mac, I set my Network Interface to 10 MBit/s and JP1 is over the pins 2 and 3 to use MII mode. The read out of the general information works but nothing is displayed when I ping the Board. 

When I hook up Wireshark to the network interface I see the ARP requests sent by my PC but no Responses from the Board. The LEDs connected to the PHY Chip indicate activity when I send out ARP requests. 

Afterwards I tried the the `eth_rd` program, again with no output. Next I tried the `eth_wr_` program and I got no UDP datagrams when hooking up wireshark to the interface. I tested `ptplib_demo` which sends out synch requests, 

Is there some further setup do?
Is there a way to debug why I don't get an output?

Edit: 
I found a bug inside the eth_wr demo program, it only writes to address 0, as the frame is only provided in binary I didn't check what if that is correct.
Additionally I found the bug inside the ethlib. The default RX Buffer descriptor is never initialised, therefore the empty bit is cleared, which indicates that the buffer is full. I added a line inside the MAC initialiation. 

Console Output of `ethlib_demo`: 
```
------------------------------------------------------------------------------
-    Demo application I for the ETH-MAC contoller and library for Patmos     -
-               Advanced Computer Architecture - Spring 2015                 -
------------------------------------------------------------------------------

Available operations:
  1 -> Print general information
  2 -> Start demo (rx, tx, LED, calculator)
  0 -> Quit

Select operation: 1

General info:
  MAC: 00:80:6E:F0:DA:42
  IP: 192.168.24.50
  UDP port for LEDS: 1234
  UDP port for calculator: 1235

ARP IP/MAC table
#	Used	IP		MAC
0	0	0.0.0.0	00:00:00:00:00:00
1	0	0.0.0.0	00:00:00:00:00:00
2	0	0.0.0.0	00:00:00:00:00:00
3	0	0.0.0.0	00:00:00:00:00:00
4	0	0.0.0.0	00:00:00:00:00:00
5	0	0.0.0.0	00:00:00:00:00:00
6	0	0.0.0.0	00:00:00:00:00:00
7	0	0.0.0.0	00:00:00:00:00:00


------------------------------------------------------------------------------
------------------------------------------------------------------------------

Available operations:
  1 -> Print general information
  2 -> Start demo (rx, tx, LED, calculator)
  0 -> Quit

Select operation: 2

Demo (rx, tx, LED, calculator)
------------------------------------------------------------------------------

Waiting to receive a packet... 
```

**Comment by egk696** [comment]
Hi Christoph, the jumper sounds correct but I've never tried 10Mbit... only 100. Do you really need 10Mbit?

**Reply by epsilon-0311** [reply]
I changed the link speed to 100 MBit no change in behavior

**Comment by epsilon-0311** [comment]
I managed to get sending of UDP packets working. After some debugging using SignalTap Logic Analyzer I found out that the demonstration program `eth_wr` does not work, as it always writes to address 0 of the buffer and does not trigger sending of the packet.  

I'm still working on checking why the reception of packets does not work.

**Reply by Emoun** [reply]
It is good that you found a problem with the demonstration program. If you have time, it would be beneficial if you could make a PR with a correction.

**Reply by epsilon-0311** [reply]
I'm currently working on finding the bug in the reception, as soon I get it running I try to also provide a corrected demonstration for that.

**Reply by epsilon-0311** [reply]
I found the bug in the ethlib, I'm testing my fix und work on the PR

**Reply by Emoun** [reply]
Very nice. Could I ask you to make an issue describing the bug and with a minimal example showing it? Then link to it in the PR that I'll gladly review.

**Reply by epsilon-0311** [reply]
Issue and PR created

**Comment by epsilon-0311** [comment, marked as answer]
PR created https://github.com/t-crest/patmos/pull/99

**Reply by schoeberl** [reply]
Is this sill relevant to be merged?

**Reply by epsilon-0311** [reply]
The PR is already merged
